hotosm/osm-stats-api
# Add TM stats Per User

**What does this PR do?**
- This PR adds stats for the `time spent mapping` and `time spent validating` per user within a given time period under the mapathon detail endpoint. 
- It also replaces the dblink functions extracting stats for the `tasks_mapped_per_user` and `tasks_validated_per_user` with direct queries from the Tasking Manager database. 
- It validates against negative tasks mapped and validated counts.

**How to test:** 
- Connect to both the Insights and TM databases. 
- Test with inputs for the mapathon detail endpoint, i.e. start date, end date, project ids, and/or hashatags via `/mapathon/detail`
- Expected response should be an object with `mappedFeatures`, `contributors`, and the `tmStats` keys. The `tmStats` contains `tasks_mapped`, `tasks_validated`, `time_spent_mapping (s)`, and `time_spent_validating (s)`stats per user for the given projects or hashtag in the given time range (that is < 24 hours).

**Example Queries**: 
1. query with project ids: `{
    "fromTimestamp": "2021-02-21T18:00:00.000",
    "toTimestamp": "2021-02-22T09:00:00.000",
    "projectIds": [
        10071,
        10334,
        9718,
        10191
    ],
    "hashtags": []
}`
2. query with TM specific hashtag: `{
    "fromTimestamp": "2021-02-21T18:00:00.000",
    "toTimestamp": "2021-02-22T09:00:00.000",
    "projectIds": [],
    "hashtags": [
        "hotosm-project-10334"
    ]
}`
3. query with non-TM specific hashtag: `{
    "fromTimestamp": "2021-02-21T18:00:00.000",
    "toTimestamp": "2021-02-22T09:00:00.000",
    "projectIds": [],
    "hashtags": [
        "missingmaps"
    ]
}`
4. query with non-existent hashtag: `{
    "fromTimestamp": "2021-02-21T18:00:00.000",
    "toTimestamp": "2021-02-22T09:00:00.000",
    "projectIds": [],
    "hashtags": [
        "artemis"
    ]
}`

**Related issues**
- Fixes #147 
- Fixes #158
- Fixes #156
- Fixes #115
